### PR TITLE
Fix class expression from being assignable if types don't match

### LINKED ISSUE #39469: Class expressions: illegal assignments are allowed
**TypeScript Version:**  4.0.0-dev.20200707

**Search Terms:** Subclassing, class expression, constructor, inheritance 

**Expected behavior:**

My anonymous class expression should not be assignable to `SortedArray` because the `brand` is missing.

**Actual behavior:**

The class expression is assignable to `SortedArray`. This compiles, but breaks in runtime.

**Related Issues:** 

**Code**

I wanted a nominal, array-like structure to represent sorted arrays. It should have the same API as `ReadonlyArray`. I'm subclassing the built-in `Array` constructor under the hood.

```ts
interface SortedArray<T> extends ReadonlyArray<T> {
  readonly brand: unique symbol; // This brand will be missing
}

interface SortedArrayConstructor {
  new <T>(items: T[], comparator: Comparator<T>): SortedArray<T>;
}

// This assignment should not be legal
const SortedArray: SortedArrayConstructor = class<T> extends Array<T> {
  constructor(items: T[], comparator: Comparator<T>) {
    super(...items.slice().sort(comparator));
  }
}
```

This breaks in runtime — we never defined `brand` on the anonymous class.

```ts
// Uncaught TypeError: Cannot read property 'toString' of undefined
new SortedArray([1, 3, 2], comparator).brand.toString()
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
const SortedArray = class extends Array {
    constructor(items, comparator) {
        super(...items.slice().sort(comparator));
    }
};
var Comparison;
(function (Comparison) {
    Comparison[Comparison["LessThan"] = -1] = "LessThan";
    Comparison[Comparison["Equal"] = 0] = "Equal";
    Comparison[Comparison["GreaterThan"] = 1] = "GreaterThan";
})(Comparison || (Comparison = {}));
const comparator = (left, right) => {
    if (left === right) {
        return Comparison.Equal;
    }
    else if (left < right) {
        return Comparison.LessThan;
    }
    else {
        return Comparison.GreaterThan;
    }
};
// Uncaught TypeError: Cannot read property 'toString' of undefined
new SortedArray([1, 3, 2], comparator).brand.toString();

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?ts=4.0.0-dev.20200707&ssl=2&ssc=32&pln=2&pc=19#)

### PR DESCRIPTION
**Pull request by tiagovtristao**
This PR aims to fix #39469

A simpler code version highlighting the problem is seen below:

```
interface A {
  prop: string;
}

const A: {new(): A} = class {}
```

The `const A` declaration should be invalid since the initializer doesn't have the `prop` property.

Currently, the initializer is having members of the `A` symbol merged into it, which is invalid in this scenario. This is however the behaviour that it is expected for `const` declarations where the initializer is function-like, as seen in https://github.com/microsoft/TypeScript/pull/26368.

### CODE REVIEW

**Review by nitish-awasthi — APPROVED** [review]

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -1966,7 +1966,7 @@ namespace ts {
         let name: Expression | BindingName | undefined;
         let decl: Node | undefined;
         if (isVariableDeclaration(node.parent) && node.parent.initializer === node) {
-            if (!isInJSFile(node) && !isVarConst(node.parent)) {
+            if (!isInJSFile(node) && !(isVarConst(node.parent) && isFunctionSymbol(node.parent.symbol))) {
```
Comment (sandersn) [inline src/compiler/utilities.ts:1969]: `isFunctionSymbol` does some redundant or unneeded work -- you can get the same effect by checking  `isFunctionLikeDeclaration(node.parent.initializer)` instead

**Review by sandersn — CHANGES_REQUESTED** [review]
I just moved getDeclarationOfExpando to getSymbolOfExpando in checker.ts. You'll need to make this change there.

File: src/compiler/utilities.ts (hunk 1 above)
Comment (tiagovtristao) [reply src/compiler/utilities.ts:1969]: Nice!

**Comment by karol-majewski** [comment]
Thank you @tiagovtristao!

**Review by sandersn — APPROVED** [review]

**Pull request merged**